From 93f4a8e3a2e5a1e286450c201aa1b1ac6e1e4a98 Mon Sep 17 00:00:00 2001
From: Claudiu Manoil <claudiu.manoil@nxp.com>
Date: Thu, 31 Aug 2017 18:08:17 +0300
Subject: [PATCH 046/706] enetc: Fix device bind/unbind dependencies

Fix double pci_disable_msix() calls.
Fix ndev dependency: int vector de-alloc after unregister_netdev(),
as unregister_netdev() calls ndo_get_stats() (still needs access to
driver priv space).

Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit 8b2306a0f5f664cc4a41814afe527e26209ff58a)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c | 12 +++---------
 1 file changed, 3 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index 11d3ec372823..8a80d2abdfd9 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -1006,12 +1006,6 @@ static void enetc_netdev_setup(struct enetc_si *si, struct net_device *ndev)
 	ndev->features = NETIF_F_HIGHDMA | NETIF_F_SG;
 }
 
-static void enetc_free_netdev(struct net_device *ndev)
-{
-	unregister_netdev(ndev);
-	free_netdev(ndev);
-}
-
 static void enetc_configure_port(struct enetc_ndev_priv *priv)
 {
 	struct enetc_hw *hw = &priv->si->hw;
@@ -1086,7 +1080,6 @@ static void enetc_free_msix(struct enetc_ndev_priv *priv)
 	}
 
 	kfree(priv->int_vector);
-	pci_disable_msix(priv->si->pdev);
 	kfree(priv->msix_entries);
 }
 
@@ -1207,10 +1200,11 @@ static void enetc_pci_remove(struct pci_dev *pdev)
 
 	dev_info(&pdev->dev, "enetc_pci_remove()\n");
 
+	unregister_netdev(si->ndev);
+
 	enetc_free_msix(netdev_priv(si->ndev));
 
-	if (si->ndev)
-		enetc_free_netdev(si->ndev);
+	free_netdev(si->ndev);
 
 	iounmap(hw->reg);
 	kfree(si);
-- 
2.17.1

